File: surf-api-bukkit/surf-api-bukkit-server/src/main/java/dev/slne/surf/surfapi/bukkit/server/libs/reflection/PaperPluginClassLoaderProxy.java
package dev.slne.surf.surfapi.bukkit.server.libs.reflection;

import dev.slne.surf.surfapi.core.api.reflection.annontation.Field;
import dev.slne.surf.surfapi.core.api.reflection.annontation.Field.Type;
import dev.slne.surf.surfapi.core.api.reflection.annontation.SurfProxy;
import io.papermc.paper.plugin.entrypoint.classloader.PaperPluginClassLoader;
import java.net.URLClassLoader;

@SurfProxy(PaperPluginClassLoader.class)
public interface PaperPluginClassLoaderProxy {

  @Field(name = "libraryLoader", type = Type.SETTER, overrideFinal = true)
  void setLibraryLoader(PaperPluginClassLoader instance, URLClassLoader libraryLoader);

  @Field(name = "libraryLoader", type = Type.GETTER)
  URLClassLoader getLibraryLoader(PaperPluginClassLoader instance);
}

File: surf-api-bukkit/surf-api-bukkit-server/src/main/kotlin/dev/slne/surf/surfapi/bukkit/server/libs/reflection/PaperPluginClassLoaderProxy.kt
package dev.slne.surf.surfapi.bukkit.server.libs.reflection

import dev.slne.surf.surfapi.core.api.reflection.Field
import dev.slne.surf.surfapi.core.api.reflection.Field.Type
import dev.slne.surf.surfapi.core.api.reflection.SurfProxy
import io.papermc.paper.plugin.entrypoint.classloader.PaperPluginClassLoader
import java.net.URLClassLoader

@SurfProxy(PaperPluginClassLoader::class)
interface PaperPluginClassLoaderProxy {
    @Field("libraryLoader", Type.SETTER, overrideFinal = true)
    fun setLibraryLoader(instance: PaperPluginClassLoader, libraryLoader: URLClassLoader)

    @Field("libraryLoader", Type.GETTER)
    fun getLibraryLoader(instance: PaperPluginClassLoader): URLClassLoader
}
